Framework/Algorithms/src/ConvertUnits.cpp

```diff
@@ -686,27 +675,22 @@ API::MatrixWorkspace_sptr ConvertUnits::removeUnphysicalBins(
                                                  maxBins - 1);
     // Next, loop again copying in the correct range for each spectrum
     for (int64_t j = 0; j < int64_t(numSpec); ++j) {
-      const MantidVec &X = workspace->readX(j);
-      const MantidVec &Y = workspace->readY(j);
-      const MantidVec &E = workspace->readE(j);
-      MantidVec &Xnew = result->dataX(j);
-      MantidVec &Ynew = result->dataY(j);
-      MantidVec &Enew = result->dataE(j);
-      int k;
-      for (k = 0; k < lastBins[j] - 1; ++k) {
-        Xnew[k] = X[k];
-        Ynew[k] = Y[k];
-        Enew[k] = E[k];
-      }
-      Xnew[k] = X[k];
-      ++k;
-      // If necessary, add on some fake values to the end of the X array (Y&E
-      // will be zero)
-      if (k < maxBins) {
-        for (int l = k; l < maxBins; ++l) {
-          Xnew[l] = X[k] + 1 + l - k;
-        }
-      }
+      auto edges = workspace->binEdges(j);
+      auto &X = workspace->x(j);
+      auto k = lastBins[j];
+
+      result->mutableX(j).assign(edges.cbegin(), edges.cbegin() + lastBins[j]);
+
+      // If the entire X range is not covered, generate fake values.
+      std::generate(result->mutableX(j).begin() + k, result->mutableX(j).end(),
+                    [ l = k, k, j, X ]() mutable {
+                      return X[k] + 1 + static_cast<double>(l++ - k);
+                    });
```

Comment:
Looking at this again, can't we write slightly simpler? Consider this `std::generate` in combination with the assign above:

```cpp
result->mutableX(j).assign(edges.cbegin(), edges.cbegin() + lastBins[j]);
```
If we just extend that until the end of the result X, you do not need to have `X[k]` in the generate.